openedx/core/djangoapps/content_libraries/models.py

```diff
@@ -127,3 +202,248 @@ def save(self, *args, **kwargs):  # lint-amnesty, pylint: disable=arguments-diff
+            email = f'{uid}@{ContentLibrariesConfig.name}'
+            with transaction.atomic():
+                if self.user is None:
+                    self.user, created = User.objects.get_or_create(
+                        username=username,
+                        defaults={'email': email})
+                    if created:
+                        # LTI users can only auth throught LTI launches.
+                        self.user.set_unusable_password()
+                    self.user.save()
+                super().save(*args, **kwds)
+
+    def __str__(self):
+        return self.subject_id
+
+
+class LtiGradedResourceManager(models.Manager):
+    """
+    A custom manager for the graded resources model.
+    """
+
+    def upsert_from_ags_launch(self, user, block, resource_endpoint, resource_link):
+        """
+        Update or create a graded resource at AGS launch.
+        """
+        resource_id = resource_link['id']
+        resource_title = resource_link.get('title') or None
+        lineitem = resource_endpoint['lineitem']
+        lti_profile = user.contentlibraries_lti_profile
+        try:
+            resource = self.get(
+                profile=lti_profile,
+                usage_key=block.scope_ids.usage_id
+            )
+        except self.model.DoesNotExist:
+            resource = self.model(
+                profile=lti_profile,
+                usage_key=block.scope_ids.usage_id,
+            )
+        resource.resource_title = resource_title
+        resource.resource_id = resource_id
+        resource.ags_lineitem = lineitem
+        resource.save()
+        return resource
+
+    def get_from_user_id(self, user_id, **kwds):
+        """
+        Retrieve a resource for a given user id holding an lti profile.
+        """
+        try:
+            user = get_user_model().objects.get(pk=user_id)
+        except get_user_model().DoesNotExist as exc:
+            raise self.model.DoesNotExist('User specified was not found.') from exc
+        profile = getattr(user, 'contentlibraries_lti_profile', None)
+        if not profile:
+            raise self.model.DoesNotExist('User does not have a LTI profile.')
+        kwds['profile'] = profile
+        return self.get(**kwds)
+
+
+class LtiGradedResource(models.Model):
+    """
+    A content libraries resource launched through LTI with AGS enabled.
+
+    Essentially, an instance of this model represents a successful LTI AGS
+    launch.  This model links the profile that launched the resource with the
+    resource itself, allowing identifcation of the link through its usage key
+    string and user id.
+    """
+
+    objects = LtiGradedResourceManager()
+
+    profile = models.ForeignKey(
+        LtiProfile,
+        on_delete=models.CASCADE,
+        related_name='lti_resources',
+        help_text=_('The authorized LTI profile that launched the resource.'))
+
+    usage_key = models.CharField(
```

Comment:
Nit: probably better to use [`UsageKeyField`](https://github.com/edx/opaque-keys/blob/46f2886d411e9cf65d25e460be89f8a70e54bf98/opaque_keys/edx/django/models.py#L198)